### PR TITLE
Benchmarks sanity checks 

### PR DESCRIPTION
**Pull request by NikVolf**
this also introduce handy externalities for state inspection, so you can do:

```rust
context.client.state_at(&BlockId::number(1)).unwrap().inspect_with(|| { ... runtime read-only code .. });
```

### CODE REVIEW

File: bin/node/bench/src/import.rs [hunk 1]
```diff
@@ -155,6 +158,41 @@ impl core::Benchmark for ImportBenchmark {
 		context.import_block(self.block.clone());
 		let elapsed = start.elapsed();
 
+		// Sanity checks.
+		context.client.state_at(&BlockId::number(1)).expect("state_at failed for block#1")
+			.inspect_with(|| {
+				match self.block_type {
+					BlockType::RandomTransfersKeepAlive => {
+						// should be 5 per signed extrinsic + 1 per unsigned
```
Comment (NikVolf) [inline bin/node/bench/src/import.rs:166]: are those sanity checks correct @shawntabrizi ?

btw, do we want to write less events on transfer?
we can merge (1, 2) events into one for sure (like `RawEvent::EndowNew`)

File: primitives/state-machine/src/read_only.rs [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+	///
+	/// Self will be set as read-only externalities and inspection
+	/// closure will be run against it.
+	fn inspect_with<F: FnOnce()>(&self, f: F);
+}
+
+impl<H: Hasher, B: Backend<H>> InspectState<H, B> for B {
+	fn inspect_with<F: FnOnce()>(&self, f: F) {
+		let mut ext = ReadOnlyExternalities::from(self);
+		ext.execute_with(f)
```
Comment (bkchr) [inline primitives/state-machine/src/read_only.rs:46]:
```suggestion
		ReadOnlyExternalities::from(self).execute_with(f)
```

File: primitives/state-machine/src/read_only.rs [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+use sp_core::{
+	storage::ChildInfo,
+	traits::Externalities, Blake2Hasher,
+};
+use codec::Encode;
+
+/// Trait for inspecting state in any backend.
+///
+/// Implemented for ay backend.
+pub trait InspectState<H: Hasher, B: Backend<H>> {
```
Comment (bkchr) [inline primitives/state-machine/src/read_only.rs:35]: Nice :+1:

**Comment by NikVolf** [comment]
> While I know that you added two benchmarks to the gitlab.yml, we may could remove them and add them as normal rust tests? Probably makes it easier to run them locally when they fail.

I added them because outside tools (i.e. bot) require executable to work when invoked with specific arguments (and in case of changes, required update to arguments would be spottable).

So maybe let's leave them?

**Comment by bkchr** [comment]
I did not proposed to remove them, I said we should call them from a Rust test, e.g.:
```
#[test]
fn test_benchmark() {
     call_simple_benchmark()
}
```

And remove the `simple_benchmark` from the gitlab.yml. But it is not that important

**Comment by NikVolf** [comment]
> I did not proposed to remove them, I said we should call them from a Rust test, e.g.:
> 
> ```
> #[test]
> fn test_benchmark() {
>      call_simple_benchmark()
> }
> ```


Yeah, I understand. I just say with those tests I want to ensure that some cli command is successful. Rust test invoking cli would be awkward and less spottable imo.

**Comment by bkchr** [comment]
> Rust test invoking cli would be awkward and less spottable imo.

That is the way for testing CLI ;) And we also do this.

**Pull request merged**